### PR TITLE
Vending machine storage - #5071

### PR DESCRIPTION
**Pull request by Fishfish458**
## About the PR 
fixes issue #5071

This PR:
Adds ability to insert items into vending machines that have a whitelist

Why isn't this a separate type?:
A functionality of ss13 was the ability for the chef and bartender to insert drinks into vending machines of that specific type and since this was along the same lines I wanted to ensure that functionality would be possible.

**Screenshots**

can insert into it

can produce

can insert food

can get random

**Changelog**

:cl:
- add: Smart fridges now properly store pills, bottles, and produce, food and drink can be added to some vending machines

### CODE REVIEW

File: Content.Server/VendingMachines/VendingMachineComponent.cs [hunk 1]
```diff
@@ -198,10 +249,56 @@ private void TryEject(string id, EntityUid? sender)
...
+            }
+            if (!_entMan.TryGetComponent(itemEntity, typeof(ItemComponent), out var item))
+            {
+                return false;
+            }
+            var metaData = _entMan.GetComponent<MetaDataComponent>(item.Owner);
+            if (metaData.EntityPrototype == null)
+            {
+                return false;
+            }
```
Comment (metalgearsloth) [inline Content.Server/VendingMachines/VendingMachineComponent.cs:293]: Functionality should never be based on a prototype's ID. It's only to be used as an identifier for spawning and nothing more. This is why entity whitelist doesn't even support it.

File: Content.Server/VendingMachines/VendingMachineComponent.cs [hunk 2]
```diff
@@ -21,13 +22,18 @@
...
+using Content.Server.Items;
+using Content.Shared.Popups;
 using static Content.Shared.Wires.SharedWiresComponent;
 
 namespace Content.Server.VendingMachines
 {
     [RegisterComponent]
     [ComponentReference(typeof(IActivate))]
-    public class VendingMachineComponent : SharedVendingMachineComponent, IActivate, IBreakAct, IWires
+    public class VendingMachineComponent : SharedVendingMachineComponent, IActivate, IBreakAct, IWires, IInteractUsing
```
Comment (metalgearsloth) [inline Content.Server/VendingMachines/VendingMachineComponent.cs:36]: No new code should be using the obsolete interfaces. IMO this should also be a separate component.

File: Content.Server/VendingMachines/VendingMachineComponent.cs [hunk 3]
```diff
@@ -198,10 +249,56 @@ private void TryEject(string id, EntityUid? sender)
...
+            }
+            if (!IsAuthorized(eventArgs.User))
+            {
+                return false;
+            }
+            if (_entMan.GetComponent<HandsComponent>(eventArgs.User).GetActiveHand?.Owner is not {Valid: true} itemEntity)
+            {
+                eventArgs.User.PopupMessage(Loc.GetString("vending-machine-component-interact-using-no-active-hand"));
+                return false;
+            }
```
Comment (metalgearsloth) [inline Content.Server/VendingMachines/VendingMachineComponent.cs:280]: Don't GetComponent if you don't know if they have the component.

File: Content.Server/VendingMachines/VendingMachineComponent.cs [hunk 4]
```diff
@@ -198,10 +249,56 @@ private void TryEject(string id, EntityUid? sender)
...
+            if (!Whitelist.IsValid(itemEntity))
+            {
+                return false;
+            }
+            if (!_entMan.TryGetComponent(itemEntity, typeof(ItemComponent), out var item))
+            {
+                return false;
+            }
```
Comment (metalgearsloth) [inline Content.Server/VendingMachines/VendingMachineComponent.cs:288]: Just use the generic version?

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Comment by MeltedPixel** [comment]
Additionally, based on your posted gifs. The items placed in the 'Smart Fridge' should stack.

I also agree with @metalgearsloth, this is no longer a vending machine at this point and should really have its own component.

**Pull request closed without merging**